doc/source/examples.rst

```diff
@@ -180,6 +180,59 @@ Unmeasured qubits are ignored by the measurement objects. Also, the
 order that qubits appear in the results is defined by the order the user added
 the measurements and not the qubit ids.
 
+The measurement gate is only used to sample the measured bitstrings and does not
+have  any effect on the state vector. If the user wishes to access the full
+state vector of a circuit that was measured, this is possible using the
+:py:attr:`qibo.base.circuit.BaseCircuit.final_state` property of circuits.
+There reasons that measurement gates do not affect the state are first that
+when more than one measurement shots are used, then the final collapsed state
+is not uniquely defined and second that the user may wish to resample the final
+state vector to obtain more shots, without having to re-execute the simulation.
+
+For applications that require the state vector to be collapsed according to a
+single-shot measurement, Qibo provides the :class:`qibo.base.gates.Collapse`
+gate. This can be used in any place within the circuit, for example:
+
+.. code-block:: python
+
+    from qibo.models import Circuit
+    from qibo import gates
+
+    c = Circuit(2)
+    c.add([gates.H(0), gates.H(1)])
+    c.add(gates.Collapse(0, result=1))
+    final_state = c()
+    # ``final_state`` will be [0, 0, 1, 1] / sqrt(2)
+
+
+Although state collapse typically happens after measurements, currently Qibo
+does not allow to add gates to qubits that are measured, and therefore
+:class:`qibo.base.gates.Collapse` cannot be used after measurements.
+An approach that allows re-using measured qubits after measuring and collapsing
+them is to construct a new circuit and pass the old final state as input:
+
+.. code-block:: python
+
+    from qibo.models import Circuit
+    from qibo import gates
+
+    c1 = Circuit(5)
+    c1.add((gates.H(i) for i in range(5)))
+    c1.add(gates.M(0, 2, 3))
+    # execute and perform a single shot measurement of qubits 0, 2 and 3
+    result = c1(nshots=1)
+
+    # create a new circuit that acts post-measurement
+    c2 = Circuit(5)
+    # take the measured bitstring from ``result``
+    bitstring = result.samples(binary=True)[0]
+    # collapse the state according to this bitstring
+    c2.add(gates.Collapse(0, 2, 3, result=bitstring))
+    # do more gates
+    c2.add((gates.H(i) for i in range(5)))
+    # execute using the measured final state of ``c1`` as the initial state
+    final_state = c2(initial_state=c1.final_state)
+
```

Comment:
The mechanism looks ok, and I believe also memory is preserved, right?

Reply:
We discussed that yesterday, but just so that we also have it written for future reference. There are two points:
* Yes, the piece of code given in this example does not create any copies of the state vector. If custom operators are used, the `c1.final_state` and the state that will be propagated in `c2` should point to the same Tensor in memory.
* The `gates.Collapse` gate, currently creates copies of the state internally because it uses `tf.reduce_sum` to calculate the norm. If we successfully move this to custom operators then we will probably manage to preserve memory completely.